tasks/build/build_action.js

```diff
@@ -1,122 +1,49 @@
-var join = require('path').join;
 var inquirer = require('inquirer');
-var execFileSync = require('child_process').execFileSync;
 
-module.exports = function (plugin) {
-  return new Promise(function (resolve, reject) {
-
-    function main() {
-      var kibanaVersion = (plugin.pkg.kibana && plugin.pkg.kibana.version) || plugin.pkg.version;
-      var deps = Object.keys(plugin.pkg.dependencies || {});
-      var buildId = `${plugin.id}-${plugin.version}`;
+var createBuild = require('./create_build');
 
-      if (kibanaVersion === 'kibana') {
-        askForKibanaVersion(function (customKibanaVersion) {
-          build(buildId, deps, customKibanaVersion);
-        });
-      } else {
-        build(buildId, deps, kibanaVersion);
-      }
-    }
+module.exports = function (plugin, run, options) {
+  options = options || {};
+  var buildVersion = plugin.version;
+  var kibanaVersion = (plugin.pkg.kibana && plugin.pkg.kibana.version) || plugin.pkg.version;
+  var buildFiles = plugin.buildSourcePatterns;
 
-    function askForKibanaVersion(cb) {
-      inquirer.prompt([
-        {
-          type: 'input',
-          name: 'kibanaVersion',
-          message: 'What version of Kibana are you building for?'
-        }
-      ]).then(function (answers) {
-        cb(answers.kibanaVersion);
-      });
-    }
+  // allow source files to be overridden
+  if (options.files && options.files.length) {
+    buildFiles = options.files;
+  }
 
-    function toBuffer(string) {
-      if (typeof Buffer.from === 'function') {
-        return Buffer.from(string, 'utf8');
-      } else {
-        // this was deprecated in node v5 in favor
-        // of Buffer.from(string, encoding)
-        return new Buffer(string, 'utf8');
-      }
-    }
+  // allow options to override plugin info
+  if (options.buildVersion) buildVersion = options.buildVersion;
+  if (options.kibanaVersion) kibanaVersion = options.kibanaVersion;
 
-    function gitInfo() {
-      try {
-        var LOG_SEPARATOR = '||';
-        var commitCount = execFileSync('git', ['rev-list', '--count', 'HEAD'], {
-          cwd: plugin.root,
-          stdio: ['ignore', 'pipe', 'ignore'],
-          encoding: 'utf8',
-        });
-        var logLine = execFileSync('git', ['log', '--pretty=%h' + LOG_SEPARATOR + '%cD', '-n', '1'], {
-          cwd: plugin.root,
-          stdio: ['ignore', 'pipe', 'ignore'],
-          encoding: 'utf8',
-        }).split(LOG_SEPARATOR);
+  // add dependency files
+  var deps = Object.keys(plugin.pkg.dependencies || {});
+  if (deps.length === 1) {
+    buildFiles.push(`node_modules/${ deps[0] }/**/*`);
+  } else if (deps.length) {
+    buildFiles.push(`node_modules/{${ deps.join(',') }}/**/*`);
+  }
 
-        return {
-          count: commitCount.trim(),
-          sha: logLine[0].trim(),
-          date: logLine[1].trim(),
-        };
-      } catch (e) {
-        return {};
-      }
+  return new Promise(function (resolve, reject) {
```

Comment:
We should either pass the reject function to the promise returned by `createBuild()`, or convert `askForKibanaVersion()` to just return it's promise and get rid of the promise constructor.